controlplane/pkg/apis/remote/connection/helpers.go

```diff
@@ -1,11 +1,134 @@
+// Copyright 2018 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package connection
 
-func (*Connection) IsValid() error {
-	// TODO
+import (
+	fmt "fmt"
+	"net"
+	"strconv"
+)
+
+// IsValid - Did you tell me enough that some other party in the chain can fill in the blanks
+func (c *Connection) IsValid() error {
+	if c == nil {
+		return fmt.Errorf("Connection cannot be nil")
+	}
+	if c.GetNetworkService() == "" {
+		return fmt.Errorf("Connection.NetworkService cannot be empty: %v", c)
+	}
+
+	if c.GetMechanism() != nil {
+		if err := c.GetMechanism().isValid(); err != nil {
+			return fmt.Errorf("Invalid Mechanism in %v: %s", c, err)
+		}
+	}
+	return nil
+}
+
+// IsComplete - Have I been told enough to actually give you what you asked for
+func (c *Connection) IsComplete() error {
+	if err := c.IsValid(); err != nil {
+		return err
+	}
+
+	if c.GetId() == "" {
+		return fmt.Errorf("Connection.Id cannot be empty: %v", c)
+	}
+
+	if c.GetContext() == nil {
+		return fmt.Errorf("Connection.Context cannot be nil: %v", c)
+	}
+
 	return nil
 }
 
-func (*Connection) IsComplete() error {
-	// TODO
+func (m *Mechanism) isValid() error {
+	if m == nil {
+		return fmt.Errorf("Mechanism cannot be nil")
+	}
+	if m.GetParameters() == nil {
+		return fmt.Errorf("Mechanism.Parameters cannot be nil: %v", m)
+	}
+
+	if m.GetType() == MechanismType_VXLAN {
+		if _, err := m.SrcIP(); err != nil {
+			return fmt.Errorf("Mechanism.Type %s requires Mechanism.Parameters[%s] for network namespace", m.GetType(), VXLANSrcIP)
+		}
+
+		if _, err := m.DstIP(); err != nil {
+			return fmt.Errorf("Mechanism.Type %s requires Mechanism.Parameters[%s] for network namespace", m.GetType(), VXLANDstIP)
+		}
+
+		if _, err := m.VNI(); err != nil {
```

Comment:
VNI's are 24 bit, we should range check the uint32 here to make sure we are actually in the valid 24 bit range :)

Reply:
m.VNI() will always return a 24-bit number or an error if it does not fit